src/test/java/hudson/plugins/build_timeout/operations/AbortAndRestartOperationTest.java

```diff
@@ -0,0 +1,157 @@
+
+import java.util.LinkedList;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import hudson.model.Cause;
+import hudson.model.FreeStyleProject;
+import hudson.model.Result;
+import hudson.plugins.build_timeout.*;
+import hudson.plugins.build_timeout.BuildTimeOutJenkinsRule;
+import hudson.plugins.build_timeout.BuildTimeoutWrapper;
+import hudson.plugins.build_timeout.impl.AbsoluteTimeOutStrategy;
+import hudson.tasks.BatchFile;
+import hudson.tasks.Shell;
+
+
+public class AbortAndRestartOperationTest {
+    
+    @Rule
+    public BuildTimeOutJenkinsRule j = new BuildTimeOutJenkinsRule();
+
+    @Test
+    public void testAbortAndRestartOnce() throws Exception {
+        
+        FreeStyleProject testproject = j.createFreeStyleProject();
+
+        AbsoluteTimeOutStrategy strategy = new AbsoluteTimeOutStrategy(Integer.toString(3)); //timeoutMinutes
+        AbortAndRestartOperation operation = new AbortAndRestartOperation(1); //Number of restarts
+        LinkedList<BuildTimeOutOperation> list = new LinkedList<BuildTimeOutOperation>();
+        list.add(operation);
+        
+        BuildTimeoutWrapper wrapper = new BuildTimeoutWrapper(strategy,list,"");
+        testproject.getBuildWrappersList().add(wrapper);
+        
+        if(System.getProperty("os.name").contains("Windows")){
+            BatchFile builder = new BatchFile("ping -n 500 127.0.0.1 &amp;gt;nul");
+            testproject.getBuildersList().add(builder);
+        }else{
+            Shell builder = new Shell("sleep 500");
+            testproject.getBuildersList().add(builder);
+        }
+        
+        assertTrue(testproject.getBuilds().size()==0);
+                
+        
+        testproject.scheduleBuild(new Cause.UserIdCause());
+        try{
+            Thread.sleep((3*60*1000)+20000);
+        }catch(InterruptedException e){
+            //Nothing todo here.
+        }
+        assertTrue(testproject.getFirstBuild() != null);
+        assertTrue(testproject.getFirstBuild().equals(testproject.getLastBuild()));
+        assertEquals(testproject.getBuilds().size(), 1);
+        
+        try{
+            Thread.sleep((3*60*1000)+20000);
+        }catch(InterruptedException e){
+            //Nothing todo here.
+        }
+        assertTrue(testproject.getFirstBuild() != null);
+        assertFalse(testproject.getFirstBuild().equals(testproject.getLastBuild()));
+        assertEquals(testproject.getBuilds().size(), 2);
+        
+        Thread.sleep(5000);
+        assertEquals(testproject.getBuilds().size(), 2);
+        assertEquals(Result.ABORTED, testproject.getFirstBuild().getResult());
+        assertEquals(Result.ABORTED, testproject.getLastBuild().getResult());
+    }
+    
+    @Test
+    public void testAbortAndRestartTwice() throws Exception {
+        
+        FreeStyleProject testproject = j.createFreeStyleProject();
+
+        AbsoluteTimeOutStrategy strategy = new AbsoluteTimeOutStrategy(Integer.toString(3)); //timeoutMinutes
```

Comment:
You can save the test execution time by using [`QuickTimeoutStrategy`](https://github.com/jenkinsci/build-timeout-plugin/blob/build-timeout-1.15/src/test/java/hudson/plugins/build_timeout/QuickBuildTimeOutStrategy.java) (available only in tests).
https://github.com/jenkinsci/build-timeout-plugin/blob/build-timeout-1.15/src/test/java/hudson/plugins/build_timeout/operations/BuildStepOperationTest.java#L142